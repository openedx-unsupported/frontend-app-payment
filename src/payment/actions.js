import { utils } from '../common';

const { AsyncActionType } = utils;

export const FETCH_BASKET = new AsyncActionType('PAYMENT', 'FETCH_BASKET');


export const fetchBasket = () => ({
  type: FETCH_BASKET.BASE,
  payload: {},
});

export const fetchBasketBegin = () => ({
  type: FETCH_BASKET.BEGIN,
});

export const fetchBasketSuccess = result => ({
  type: FETCH_BASKET.SUCCESS,
  payload: result,
});

export const fetchBasketFailure = () => ({
  type: FETCH_BASKET.FAILURE,
});
