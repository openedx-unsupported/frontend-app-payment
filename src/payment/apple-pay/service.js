/* eslint-disable no-console */
import pick from 'lodash.pick';
import { applyConfiguration } from '../../common/serviceUtils';

let apiClient = null;

let config = {
  ECOMMERCE_BASE_URL: null,
  ECOMMERCE_RECEIPT_BASE_URL: null,
  APPLE_PAY_MERCHANT_IDENTIFIER: null,
  APPLE_PAY_MERCHANT_NAME: null,
  APPLE_PAY_COUNTRY_CODE: null,
  APPLE_PAY_CURRENCY_CODE: null,
  APPLE_PAY_START_SESSION_URL: null,
  APPLE_PAY_AUTHORIZE_URL: null,
  APPLE_PAY_SUPPORTED_NETWORKS: null,
  APPLE_PAY_MERCHANT_CAPABILITIES: null,
};

export function configureApiService(_config, _apiClient) {
  applyConfiguration(config, _config);
  config = pick(_config, Object.keys(config));
  apiClient = _apiClient;
}

/* istanbul ignore next */
const tempLogAxiosError = (error) => {
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    console.log(error.response.data);
    console.log(error.response.data.statusMessage);
    console.log(error.response.status);
    console.log(error.response.headers);
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    console.log(error.request);
  } else {
    // Something happened in setting up the request that triggered an Error
    console.log('Error', error.message);
  }
  console.log(error.config);
};


export const performApplePayPayment = ({
  totalAmount,
  onPaymentBegin,
  onPaymentComplete,
  onPaymentAuthorizationFailure,
  onMerchantValidationFailure,
  onPaymentCancel,
}) => {
  // Session Set Up
  // ------------------------

  const version = 2;
  const paymentRequest = {
    countryCode: config.APPLE_PAY_COUNTRY_CODE,
    currencyCode: config.APPLE_PAY_CURRENCY_CODE,
    supportedNetworks: config.APPLE_PAY_SUPPORTED_NETWORKS,
    merchantCapabilities: config.APPLE_PAY_MERCHANT_CAPABILITIES,
    total: {
      label: config.APPLE_PAY_MERCHANT_NAME,
      type: 'final',
      amount: totalAmount,
    },
    requiredBillingContactFields: ['postalAddress'],
  };

  const applePaySession = new global.ApplePaySession(version, paymentRequest);


  // Set Up Event Handlers on Payment Session
  // ------------------------

  applePaySession.onvalidatemerchant = (event) => {
    console.log('Validating merchant...');

    const postData = {
      url: event.validationURL,
      is_payment_microfrontend: true,
    };

    return apiClient.post(config.APPLE_PAY_START_SESSION_URL, postData)
      .then((response) => {
        console.log('Merchant validation succeeded.', response.data);
        applePaySession.completeMerchantValidation(response.data);
      })
      .catch((error) => {
        tempLogAxiosError(error);
        applePaySession.abort();
        /* istanbul ignore else */
        if (onMerchantValidationFailure) onMerchantValidationFailure(error);
      });
  };

  applePaySession.onpaymentauthorized = (event) => {
    console.log('Submitting Apple Pay payment to CyberSource...');

    const postData = event.payment;

    return apiClient.post(config.APPLE_PAY_AUTHORIZE_URL, postData)
      .then(({ data }) => {
        const orderNumber = data.number;
        console.log('CyberSource successfully authorized Apple Pay payment.', data);
        applePaySession.completePayment(global.ApplePaySession.STATUS_SUCCESS);
        /* istanbul ignore else */
        if (onPaymentComplete) onPaymentComplete(orderNumber, true);
      })
      .catch((error) => {
        tempLogAxiosError(error);
        applePaySession.completePayment(global.ApplePaySession.STATUS_FAILURE);
        /* istanbul ignore else */
        if (onPaymentAuthorizationFailure) onPaymentAuthorizationFailure(error);
      });
  };

  applePaySession.oncancel = (event) => {
    console.log('Cancel', event);
    /* istanbul ignore else */
    if (onPaymentCancel) onPaymentCancel(event);
  };


  // Begin transaction
  // ------------------------
  applePaySession.begin();
  if (onPaymentBegin) onPaymentBegin();
};

/**
 * Given an order number, and an optional arg of whether to disable the back button, redirects the
 * user to the receipt page with the correct query parameters.
 *
 * @param orderNumber The number of the order
 * @param disableBackButton Whether to disable the back button on the receipt page. We default to
 * false since the receipt page is referenced in emails/etc. We only disable the back button when
 * the flow goes through the payment page.
 */
/* istanbul ignore next - you can't mock window.location.href in jsdom */
export const redirectToReceipt = (orderNumber, disableBackButton = false) => {
  const queryParams = `order_number=${orderNumber}&disable_back_button=${Number(disableBackButton)}`;
  global.location.href = `${config.ECOMMERCE_RECEIPT_BASE_URL}?${queryParams}`;
};
