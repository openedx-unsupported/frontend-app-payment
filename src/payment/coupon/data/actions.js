import { AsyncActionType } from '../../../common/utils';

export const ADD_COUPON = new AsyncActionType('PAYMENT', 'ADD_COUPON');
export const REMOVE_COUPON = new AsyncActionType('PAYMENT', 'REMOVE_COUPON');
export const UPDATE_COUPON_DRAFT = 'PAYMENT__UPDATE_COUPON_DRAFT';

// ADD COUPON

export const addCoupon = code => ({
  type: ADD_COUPON.BASE,
  payload: {
    code,
  },
});

export const addCouponBegin = () => ({
  type: ADD_COUPON.BEGIN,
});

export const addCouponSuccess = (voucherId, code, benefit) => ({
  type: ADD_COUPON.SUCCESS,
  payload: {
    benefit,
    code,
    voucherId,
  },
});

export const addCouponFailure = errorCode => ({
  type: ADD_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// REMOVE COUPON

export const removeCoupon = voucherId => ({
  type: REMOVE_COUPON.BASE,
  payload: {
    voucherId,
  },
});

export const removeCouponBegin = () => ({
  type: REMOVE_COUPON.BEGIN,
});

export const removeCouponSuccess = result => ({
  type: REMOVE_COUPON.SUCCESS,
  payload: {
    result,
  },
});

export const removeCouponFailure = errorCode => ({
  type: REMOVE_COUPON.FAILURE,
  payload: {
    errorCode,
  },
});

// UPDATE COUPON DRAFT

export const updateCouponDraft = code => ({
  type: UPDATE_COUPON_DRAFT,
  payload: {
    code,
  },
});
