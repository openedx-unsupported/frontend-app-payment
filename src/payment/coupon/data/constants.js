export const PERCENTAGE_BENEFIT = 'Percentage';
export const ABSOLUTE_BENEFIT = 'Absolute';
