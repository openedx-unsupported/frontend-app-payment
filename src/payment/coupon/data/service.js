import pick from 'lodash.pick';

import { handleRequestError, applyConfiguration } from '../../../common/serviceUtils';

let config = {
  ECOMMERCE_BASE_URL: null,
};

let apiClient = null; // eslint-disable-line no-unused-vars

export function configureApiService(newConfig, newApiClient) {
  applyConfiguration(config, newConfig);
  config = pick(newConfig, Object.keys(config));
  apiClient = newApiClient;
}

export async function postCoupon(code) {
  const response = await apiClient
    .post(
      `${config.ECOMMERCE_BASE_URL}/payment-bff/v0/payment/vouchers/`,
      { code },
      {
        headers: { 'Content-Type': 'application/json' },
      },
    )
    .catch(handleRequestError);

  return response.data;
}

export async function deleteCoupon(voucherId) {
  const response = await apiClient
    .delete(`${config.ECOMMERCE_BASE_URL}/payment-bff/v0/payment/vouchers/${voucherId}`)
    .catch(handleRequestError);

  return response.data;
}
