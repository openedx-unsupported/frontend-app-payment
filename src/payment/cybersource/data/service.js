import formurlencoded from 'form-urlencoded';
import pick from 'lodash.pick';

import { applyConfiguration } from '../../../common/serviceUtils';


let config = {
  ECOMMERCE_BASE_URL: null,
};

let apiClient = null; // eslint-disable-line no-unused-vars

export function configureApiService(newConfig, newApiClient) {
  applyConfiguration(config, newConfig);
  config = pick(newConfig, Object.keys(config));
  apiClient = newApiClient;
}

export async function sdnCheck(firstName, lastName, city, country) {
  const { data } = await apiClient.post(
    `${config.ECOMMERCE_BASE_URL}/api/v2/sdn/search/`,
    {
      name: `${firstName} ${lastName}`,
      city,
      country,
    },
  );

  return data;
}

export async function checkout(basketId, cardHolderInfo) {
  const { data } = await apiClient.post(
    `${config.ECOMMERCE_BASE_URL}/payment/cybersource/api-submit/`,
    formurlencoded({
      basket: basketId,
      first_name: cardHolderInfo.firstName,
      last_name: cardHolderInfo.lastName,
      address_line1: cardHolderInfo.address,
      address_line2: cardHolderInfo.unit,
      city: cardHolderInfo.city,
      country: cardHolderInfo.country,
      state: cardHolderInfo.state,
      postal_code: cardHolderInfo.postalCode,
      organization: cardHolderInfo.organization,
    }),
    {
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
      },
    },
  );

  return data;
}
