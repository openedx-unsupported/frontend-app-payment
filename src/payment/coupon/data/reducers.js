import { ADD_COUPON, REMOVE_COUPON, UPDATE_COUPON_DRAFT } from './actions';

const defaultState = {
  benefit: null,
  code: null,
  voucherId: null,
  error: null,
  loaded: false,
  loading: false,
};

const reducer = (state = defaultState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case ADD_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case ADD_COUPON.SUCCESS:
        return {
          ...state,
          loading: false,
          loaded: true,
          error: null,
          benefit: action.payload.benefit,
          code: action.payload.code,
          voucherId: action.payload.voucherId,
        };
      case ADD_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: false,
          error: action.payload.error,
          voucherId: null,
        };
      case REMOVE_COUPON.BEGIN:
        return {
          ...state,
          loading: true,
          loaded: false,
          error: null,
        };
      case REMOVE_COUPON.SUCCESS:
        return defaultState;
      case REMOVE_COUPON.FAILURE:
        return {
          ...state,
          loading: false,
          loaded: true, // There's still a loaded coupon.
          error: action.payload.error,
        };
      case UPDATE_COUPON_DRAFT:
        return {
          ...state,
          code: action.payload.code,
        };
      default:
    }
  }

  return state;
};

export default reducer;
