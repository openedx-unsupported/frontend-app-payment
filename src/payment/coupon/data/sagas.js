import { call, put, takeEvery } from 'redux-saga/effects';
import {
  ADD_COUPON,
  addCouponBegin,
  addCouponSuccess,
  addCouponFailure,
  removeCouponBegin,
  removeCouponSuccess,
  removeCouponFailure,
  REMOVE_COUPON,
} from './actions';
import { postCoupon } from './service';

function* handleAddCoupon(action) {
  yield put(addCouponBegin());
  try {
    const result = yield call(postCoupon, action.payload.code);
    const { id: voucherId, code, benefit } = result.voucher;
    yield put(addCouponSuccess(voucherId, code, benefit));
  } catch (e) {
    if (e.code !== undefined) {
      yield put(addCouponFailure(e.code));
    } else {
      throw e;
    }
  }
}

function* handleRemoveCoupon(action) {
  yield put(removeCouponBegin());
  try {
    const result = yield call(postCoupon, action.payload.codel);
    yield put(removeCouponSuccess(result));
  } catch (e) {
    if (e.code !== undefined) {
      removeCouponFailure(e.code);
    } else {
      throw e;
    }
  }
}

export default function* saga() {
  yield takeEvery(ADD_COUPON.BASE, handleAddCoupon);
  yield takeEvery(REMOVE_COUPON.BASE, handleRemoveCoupon);
}
