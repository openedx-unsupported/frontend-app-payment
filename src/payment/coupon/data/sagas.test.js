import { runSaga } from 'redux-saga';

import { configureApiService } from './service';
import saga, { handleAddCoupon, handleRemoveCoupon } from './sagas';
import {
  addCoupon,
  addCouponBegin,
  addCouponSuccess,
  addCouponFailure,
  removeCouponBegin,
  removeCouponSuccess,
  removeCoupon,
  removeCouponFailure,
} from './actions';
import { fetchBasketSuccess } from '../../data/actions';
import { transformResults } from '../../data/service';
import { addMessage, INFO, DANGER } from '../../../feedback';

describe('saga tests', () => {
  const configuration = {
    ECOMMERCE_BASE_URL: 'http://localhost',
  };

  describe('main saga', () => {
    it('should setup its sub sagas correctly', () => {
      const gen = saga();

      expect(gen.next()).toMatchSnapshot();
      expect(gen.next()).toMatchSnapshot();
      expect(gen.next()).toEqual({ done: true, value: undefined });
    });
  });

  describe('handleAddCoupon', () => {
    const responses = {
      successResponse: {
        data: {
          show_coupon_form: true,
          coupons: [
            {
              id: 12345,
              code: 'DEMO25',
              benefit_value: '25%',
            },
          ],
          summary_price: 161,
          order_total: 149,
          summary_discounts: 12,
          products: [
            {
              image_url: 'https://prod-discovery.edx-cdn.org/media/course/image/21be6203-b140-422c-9233-a1dc278d7266-941abf27df4d.small.jpg',
              title: 'Introduction to Happiness',
              certificate_type: 'verified',
              product_type: 'Seat',
              sku: '8CF08E5',
            },
          ],
        },
      },
      blankVoucherResponse: {
        data: {
          coupons: [],
          show_coupon_form: true,
          summary_price: 161,
          order_total: 161,
          summary_discounts: 0,
          products: [
            {
              image_url: 'https://prod-discovery.edx-cdn.org/media/course/image/21be6203-b140-422c-9233-a1dc278d7266-941abf27df4d.small.jpg',
              title: 'Introduction to Happiness',
              certificate_type: 'verified',
              product_type: 'Seat',
              sku: '8CF08E5',
            },
          ],
        },
      },
      errorResponse: {
        response: {
          data: {
            error_code: 'uhoh',
          },
        },
      },
      unexpectedErrorResponse: 'oh snap!', // We don't know how to handle this
    };

    it('should handle success', async () => {
      const apiClientPost = jest.fn(() =>
        new Promise((resolve) => {
          resolve(responses.successResponse);
        }));

      configureApiService(configuration, {
        post: apiClientPost,
      });

      const dispatched = [];
      await runSaga(
        {
          dispatch: action => dispatched.push(action),
        },
        handleAddCoupon,
        addCoupon('DEMO25'),
      ).toPromise();

      expect(dispatched).toEqual([
        addCouponBegin(),
        fetchBasketSuccess(transformResults(responses.successResponse.data)),
        addCouponSuccess(12345, 'DEMO25', '25%'),
        addMessage(
          'payment.coupon.added',
          null,
          {
            code: 'DEMO25',
          },
          INFO,
        ),
      ]);
      expect(apiClientPost).toHaveBeenCalledWith(
        'http://localhost/bff/payment/v0/vouchers/',
        { code: 'DEMO25' },
        { headers: { 'Content-Type': 'application/json' } },
      );
    });

    it('should handle an empty coupons', async () => {
      const apiClientPost = jest.fn(() =>
        new Promise((resolve) => {
          resolve(responses.blankVoucherResponse);
        }));

      configureApiService(configuration, {
        post: apiClientPost,
      });

      const dispatched = [];
      await runSaga(
        {
          dispatch: action => dispatched.push(action),
        },
        handleAddCoupon,
        addCoupon('DEMO25'),
      ).toPromise();

      expect(dispatched).toEqual([
        addCouponBegin(),
        fetchBasketSuccess(transformResults(responses.blankVoucherResponse.data)),
        addCouponSuccess(null, null, null),
      ]);
      expect(apiClientPost).toHaveBeenCalledWith(
        'http://localhost/bff/payment/v0/vouchers/',
        { code: 'DEMO25' },
        { headers: { 'Content-Type': 'application/json' } },
      );
    });

    it('should handle an API error', async () => {
      const apiClientPost = jest.fn(() =>
        new Promise((resolve, reject) => {
          reject(responses.errorResponse);
        }));

      configureApiService(configuration, {
        post: apiClientPost,
      });

      const dispatched = [];
      await runSaga(
        {
          dispatch: action => dispatched.push(action),
        },
        handleAddCoupon,
        addCoupon('DEMO25'),
      ).toPromise();

      expect(dispatched).toEqual([
        addCouponBegin(),
        addCouponFailure(),
        addMessage(
          'uhoh',
          null,
          null,
          DANGER,
        ),
      ]);

      expect(apiClientPost).toHaveBeenCalledWith(
        'http://localhost/bff/payment/v0/vouchers/',
        { code: 'DEMO25' },
        { headers: { 'Content-Type': 'application/json' } },
      );
    });

    it('should handle an unexpected error', async () => {
      const apiClientPost = jest.fn(() =>
        new Promise((resolve, reject) => {
          reject(responses.unexpectedErrorResponse);
        }));

      configureApiService(configuration, {
        post: apiClientPost,
      });

      const caughtErrors = [];
      const dispatched = [];
      try {
        await runSaga(
          {
            dispatch: action => dispatched.push(action),
            onError: error => caughtErrors.push(error),
          },
          handleAddCoupon,
          addCoupon('DEMO25'),
        ).toPromise();
      } catch (e) {
        expect(e).toEqual('oh snap!');
      }

      expect(caughtErrors).toEqual(['oh snap!']);

      expect(apiClientPost).toHaveBeenCalledWith(
        'http://localhost/bff/payment/v0/vouchers/',
        { code: 'DEMO25' },
        { headers: { 'Content-Type': 'application/json' } },
      );
    });
  });

  describe('handleRemoveCoupon', () => {
    const responses = {
      successResponse: {
        data: {
          show_coupon_form: true,
          coupons: [
            {
              id: 12345,
              code: 'DEMO25',
              benefit_value: '25%',
            },
          ],
          summary_price: 161,
          order_total: 149,
          summary_discounts: 12,
          products: [
            {
              image_url: 'https://prod-discovery.edx-cdn.org/media/course/image/21be6203-b140-422c-9233-a1dc278d7266-941abf27df4d.small.jpg',
              title: 'Introduction to Happiness',
              certificate_type: 'verified',
              product_type: 'Seat',
              sku: '8CF08E5',
            },
          ],
        },
      },
      errorResponse: {
        response: {
          data: {
            error_code: 'uhoh',
          },
        },
      },
      unexpectedErrorResponse: 'oh snap!', // We don't know how to handle this
    };

    it('should handle success', async () => {
      const apiClientDelete = jest.fn(() =>
        new Promise((resolve) => {
          resolve(responses.successResponse);
        }));

      configureApiService(configuration, {
        delete: apiClientDelete,
      });

      const dispatched = [];
      await runSaga(
        {
          dispatch: action => dispatched.push(action),
          getState: () => ({
            payment: { coupon: { code: 'DEMO25' } },
          }),
        },
        handleRemoveCoupon,
        removeCoupon(12345),
      ).toPromise();

      expect(dispatched).toEqual([
        removeCouponBegin(),
        fetchBasketSuccess(transformResults(responses.successResponse.data)),
        removeCouponSuccess(transformResults(responses.successResponse.data)),
        addMessage(
          'payment.coupon.removed',
          null,
          {
            code: 'DEMO25',
          },
          INFO,
        ),
      ]);

      expect(apiClientDelete).toHaveBeenCalledWith('http://localhost/bff/payment/v0/vouchers/12345');
    });

    it('should handle an API error', async () => {
      const apiClientDelete = jest.fn(() =>
        new Promise((resolve, reject) => {
          reject(responses.errorResponse);
        }));

      configureApiService(configuration, {
        delete: apiClientDelete,
      });

      const dispatched = [];
      await runSaga(
        {
          dispatch: action => dispatched.push(action),
          getState: () => ({
            payment: { coupon: { code: 'DEMO25' } },
          }),
        },
        handleRemoveCoupon,
        removeCoupon(12345),
      ).toPromise();

      expect(dispatched).toEqual([
        removeCouponBegin(),
        removeCouponFailure(),
        addMessage(
          'uhoh',
          null,
          null,
          DANGER,
        ),
      ]);

      expect(apiClientDelete).toHaveBeenCalledWith('http://localhost/bff/payment/v0/vouchers/12345');
    });

    it('should handle an unexpected error', async () => {
      const apiClientDelete = jest.fn(() =>
        new Promise((resolve, reject) => {
          reject(responses.unexpectedErrorResponse);
        }));

      configureApiService(configuration, {
        delete: apiClientDelete,
      });

      const caughtErrors = [];
      const dispatched = [];
      try {
        await runSaga(
          {
            dispatch: action => dispatched.push(action),
            onError: error => caughtErrors.push(error),
            getState: () => ({
              payment: { coupon: { code: 'DEMO25' } },
            }),
          },
          handleRemoveCoupon,
          removeCoupon(12345),
        ).toPromise();
      } catch (e) {
        expect(e).toEqual('oh snap!');
      }

      expect(caughtErrors).toEqual(['oh snap!']);

      expect(apiClientDelete).toHaveBeenCalledWith('http://localhost/bff/payment/v0/vouchers/12345');
    });
  });
});
