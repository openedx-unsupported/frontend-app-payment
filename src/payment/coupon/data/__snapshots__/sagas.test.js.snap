// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`saga tests main saga should setup its sub sagas correctly 1`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "combinator": false,
    "payload": Object {
      "args": Array [
        "PAYMENT__ADD_COUPON",
        [Function],
      ],
      "context": null,
      "fn": [Function],
    },
    "type": "FORK",
  },
}
`;

exports[`saga tests main saga should setup its sub sagas correctly 2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "combinator": false,
    "payload": Object {
      "args": Array [
        "PAYMENT__REMOVE_COUPON",
        [Function],
      ],
      "context": null,
      "fn": [Function],
    },
    "type": "FORK",
  },
}
`;
